src/web/rustdoc.rs

```diff
@@ -377,18 +353,33 @@ pub fn rustdoc_html_server_handler(req: &mut Request) -> IronResult<Response> {
         inner_path.join("/")
     };
 
+    rendering_time.step("rewrite html");
+
+    let file_content = ctry!(req, std::str::from_utf8(&file.0.content));
+    let templates = req
+        .extensions
+        .get::<super::TemplateData>()
+        .expect("missing TemplateData from the request extensions");
     // Build the page of documentation
-    RustdocPage {
-        latest_path,
-        latest_version,
-        inner_path,
-        is_latest_version,
-        rustdoc_head,
-        rustdoc_body,
-        rustdoc_body_class,
-        krate,
-    }
-    .into_response(req)
+    let ctx = ctry!(
+        req,
+        tera::Context::from_serialize(RustdocPage {
+            latest_path,
+            latest_version,
+            inner_path,
+            is_latest_version,
+            krate,
+        })
+    );
+    // Extract the head and body of the rustdoc file so that we can insert it into our own html
+    let html = ctry!(req, utils::rewrite_lol(file_content, ctx, templates));
+
+    rendering_time.step("serve html");
+    use iron::{headers::ContentType, status::Status};
+    let mut response = Response::with((Status::Ok, html));
+    response.headers.set(ContentType::html());
+
+    Ok(response)
```

Comment:
Could most of this be abstracted out into a struct, I think it'd make the code much cleaner

Reply:
If you mean `impl_webpage!`, I don't see how. `WebPage` runs templating _after_ everything else, but this runs it _before_. So the only way to make it work is to have every `impl WebPage` take a custom callback, which seems like overkill if it's only going to be used here.

Reply:
Nah, not an `impl_webpage!`, that'd require reworking how the rendering works. I just mean a straight-up normal struct to make things a little cleaner. (As a side note, the imports from `iron` can probably just go at the top of the page)

Reply:
I can make it a separate function if you think that would make it easier to follow. I don't see a point in having a struct that we make exactly once just to immediately call a method on.

Reply:
err actually I say that but this would look really nice on `RustdocPage` :laughing:

Reply:
Done.